src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl

```diff
@@ -0,0 +1,195 @@
+`denomination` contains the "type" of generalized vertical coordinate (the only one implemented is `ZStar`)
+"""
+abstract type AbstractVerticalSpacing{R} end
+
+const AbstractVerticalSpacingRG  = RectilinearGrid{<:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:AbstractVerticalSpacing} 
+const AbstractVerticalSpacingLLG = LatitudeLongitudeGrid{<:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:AbstractVerticalSpacing}
+
+const AbstractVerticalSpacingUnderlyingGrid = Union{AbstractVerticalSpacingRG, AbstractVerticalSpacingLLG}
+const AbstractVerticalSpacingImmersedGrid   = ImmersedBoundaryGrid{<:Any, <:Any, <:Any, <:Any, <:AbstractVerticalSpacingUnderlyingGrid}
+
+const AbstractVerticalSpacingGrid = Union{AbstractVerticalSpacingUnderlyingGrid, AbstractVerticalSpacingImmersedGrid} 
+
+#####
+##### Original grid
+#####
+
+retrieve_static_grid(grid) = grid
+
+function retrieve_static_grid(grid::AbstractVerticalSpacingImmersedGrid) 
+    underlying_grid  = retrieve_static_grid(grid.underlying_grid)
+    active_cells_map = !isnothing(grid.interior_active_cells)
+    return ImmersedBoundaryGrid(underlying_grid, grid.immersed_boundary; active_cells_map)
+end
+
+reference_Δzᵃᵃᶠ(grid) = grid.Δzᵃᵃᶠ
+reference_Δzᵃᵃᶜ(grid) = grid.Δzᵃᵃᶜ
+
+function retrieve_static_grid(grid::AbstractVerticalSpacingUnderlyingGrid) 
+
+    Δzᵃᵃᶠ = reference_Δzᵃᵃᶠ(grid)
+    Δzᵃᵃᶜ = reference_Δzᵃᵃᶜ(grid)
+
+    TX, TY, TZ = topology(grid)
+
+    args = []
+    for prop in propertynames(grid)
+        if prop == :Δzᵃᵃᶠ
+            push!(args, Δzᵃᵃᶠ)
+        elseif prop == :Δzᵃᵃᶜ
+            push!(args, Δzᵃᵃᶜ)
+        else
+            push!(args, getproperty(grid, prop))
+        end
+    end
+
+    GridType = getnamewrapper(grid)
+
+    return GridType{TX, TY, TZ}(args...)
+end
+
+#####
+##### Some extensions
+#####
+
+function with_halo(new_halo, grid::AbstractVerticalSpacingUnderlyingGrid)
+    old_static_grid = retrieve_static_grid(grid)
+    new_static_grid = with_halo(new_halo, old_static_grid)
+    vertical_coordinate = denomination(grid)
+    new_grid = generalized_spacing_grid(new_static_grid, vertical_coordinate)
+
+    return new_grid
+end
+
+function with_halo(new_halo, ibg::AbstractVerticalSpacingImmersedGrid) 
+    underlying_grid = ibg.underlying_grid
+    immersed_boundary = ibg.immersed_boundary
+    new_underlying_grid = with_halo(new_halo, underlying_grid)
+    active_cells_map = isnothing(ibg.interior_active_cells)
+    new_ibg = ImmersedBoundaryGrid(new_underlying_grid, immersed_boundary; active_cells_map)
+    return new_ibg
+end
+
+#####
+##### General implementation
+#####
+
+generalized_spacing_grid(grid, coord) = grid
+update_vertical_spacing!(model, grid; kwargs...) = nothing
+
+##### 
+##### Vertical spacings for a generalized vertical coordinate system
+#####
+
+# Very bad for GPU performance!!! (z-values are not coalesced in memory for z-derivatives anymore)
+# TODO: make z-direction local in memory by not using Fields
```

Comment:
This comment makes me worried. There are better ways to make the z-direction local; the `Field` abstraction has nothing to do with it. Remember that `getindex` is divorced from the memory layout. The 3rd index can be local. The thing that has to change is simply how the arrays are constructed via `newdata` in the grids module.

Reply:
this is actually superseded by the fact that we do not allocate anymore a 3d field but compute \Delta z locally by doing \Delta r * s where s is a two-dimensional field, so memory access should be coalesced now